docs/cluster_api.md

```diff
@@ -0,0 +1,95 @@
+# Cluster Operator API
+
+## Overview
+
+The cluster operator API defines a set of resources that allow the instantiation of OpenShift clusters and subsequent
+updates to those clusters based on changes to said resources.
+
+The general approach for these resources is to follow the Kubernetes model of a spec section that is end-user controlled and
+a status section that represents the current state of the resources. Changes to the spec will result in the necessary actions
+by the system to reconcile the current state of the system with what the user has specified.
+
+## Resources
+
+### Cluster
+
+The cluster resource should allow 2 levels/scopes of spec:
+- Cluster-wide configuration
+- NodeGroup configuration
+
+Furthermore, each scope can be described in 2 parts:
+- Hardware spec (needed to instantiate the proper virtual hardware)
+- OpenShift config (openshift-specific configuration)
+
+For the Status part, if the cluster has been instantiated, we need to know how to access it (Kubeconfig may be sufficient)
+
+In more detail:
+
+#### Spec:
+* Cluster-wide configuration:
+  * ObjectMetadata (inline ObjectMetadata)
+  * TypeMetadata (inline TypeMetadata)
+  * Hardware-spec
+    * Type (AWS, Azure, GCE, etc) [string]
+    * AWSConfig: [struct]
+      - AWSSecret [LocalReference]
+      - Region [string]
+      - VPCName [string]
+      - VPCSubnet [string]
+  * OpenShift Configuration:
+    * DeploymentType [string]
+    * OpenShiftVersion [string]
+    * Public hostname [string]
+    * SDN Plugin Name [string]
+    * Service Network CIDR [string]
+    * Pod Network CIDR [string]
+    * Variables [array of Variable struct]
+* NodeGroups [array of NodeGroupSpec]:
+  - Type: (Master, Compute) [string]
+  - Infra: [boolean]
+  - Count: [usigned int]
+  - Hardware spec (cloud-specific):
+    - AWSConfig:
+      - InstanceType [string]
+      - AMIName [string]
+  - OpenShift Configuration:
+    - NodeLabels [array of String]
+
+#### Status:
+* Access
+  - KubeConfigSecret [LocalReference]
+* Conditions [array]
+  - Type (HardwareReady, MasterReady, etc.) [string]
+  - Status (True/False/Unknown) [string]
+  - LastProbeTime [string]
+  - LastTransitionTime [string]
+  - Message [string]
+  - Reason [string]
+* NodeGroupCount [unsigned int]
+* Provisioned [boolean]
+* Running [boolean]
+
+
+### NodeGroup
+
+#### Spec
+- Type (Master/Compute) [string]
+- Infra [boolean]
+- Count [unsigned int]
+- Hardware spec (cloud-specific):
+  - AWSConfig:
+    - InstanceType [string]
+    - AMIName [string]
```

Comment:
related to multi-AZ/VPCs/subnets, will need a field here to hold the subnet the NodeGroup will be a part of

Reply:
Thx, will add

Reply:
Who is going to update the AMI name?  What sort of user will be setting the AMI field besides the deployment automation?  How does a user look up the correct AMI name?

Reply:
@smarterclayton we need to have a mechanism to discover it, whether it's based on querying AWS for AMI tags or using a configmap. Are you saying that because it's not something end-users would know, that we should not expose it?

Reply:
I should either be setting v3.7.9 on the master and v3.7.6 on the node, or I should be setting direct AMI values and direct binaries, but not both.  This API will be driven from automation that creates these artifacts and then expects to spin up a cluster with them - whether that's a "Version" API object that you always reference by name that contains the artifacts, or a struct that takes a version ref or concrete value depending on what sort of use case it is.

Humans selecting that value aren't going to get the right AMI, and machines  probably want to be direct.

Reply:
ummmm, how can you define a nodegroup without defining the AMI?

I also don't believe one would ever update the AMI of a nodegroup. They would make a new node group and roll between them. Updating the AMI is like updating the image in a replicaset. You don't. You update the deployment and the deployment makes a new replicaset.

Reply:
Sounds like clayton wants an image stream for AMIs. In any case, it's a higher level construct then belongs in the NodeGroup.spec.  NodeGroup.spec should be a very specific AMI on AWS.

Reply:
Note that other cloud providers don't have regional AMIs (the same way), so there will be AWS specific nuances that can be ignored.

Yes, it is analogous to image stream, but it's more like

```
type Version struct {
  TypeMeta
  ObjectMeta // name is v3.7.9

  // other human presented metadata

  RPMRepos []string // populated 
  ImageFormat string // "openshift/origin-${component}:v3.7.9", or "ci.openshift.org/ci-jobs/pr-2034834098243:${component}"
  VMImages struct {
    GCPImage string
    GCPImageProject string
    
    AWSImageByRegion map[string]string
    
    ... // other cloud providers
  }
}
```